core/access/chain_access.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package access provides a layer to handle local blockchain database and
+// on-demand network retrieval
+package access
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"golang.org/x/net/context"
+)
+
+var (
+	errNotInDb = errors.New("object not found in database")
+)
+
+const LogLevel = logger.Debug
+
+var (
+	requestTimeout = time.Millisecond * 300
+	retryPeers     = time.Second * 1
+)
+
+// ChainAccess provides access to blockchain and state data through local
+// database and optionally also on-demand network retrieval
+type ChainAccess struct {
+	db         ethdb.Database
+	odr        bool // light client mode, odr enabled
+	lock       sync.Mutex
+	sentReqs   map[uint64]*sentReq
+	sentReqCnt uint64
+	peers      *peerSet
+}
+
+// requestFunc is a function that requests some data from a peer
+type requestFunc func(*Peer) error
+
+// validatorFunc is a function that processes a message and returns true if
+// it was a meaningful answer to a given request
+type validatorFunc func(*Msg) bool
+
+// sentReq is a request waiting for an answer that satisfies its valFunc
+type sentReq struct {
+	valFunc     validatorFunc
+	deliverChan chan *Msg
+}
+
+// NewDbChainAccess creates a ChainAccess with ODR disabled
+func NewDbChainAccess(db ethdb.Database) *ChainAccess {
+	return NewChainAccess(db, false)
+}
+
+// NewChainAccess create a ChainAccess with optional ODR
+func NewChainAccess(db ethdb.Database, odr bool) *ChainAccess {
+	return &ChainAccess{
+		db:       db,
+		peers:    newPeerSet(),
+		sentReqs: make(map[uint64]*sentReq),
+		odr:      odr,
+	}
+}
+
+// Db returns the local database assigned to the ChainAccess object
+func (self *ChainAccess) Db() ethdb.Database {
+	return self.db
+}
```

Comment:
Please rename the method to Database. `db` as a field name is fine, but method names should not use such drastic abbreviations.